README.md

## HeyPublisher Submisson Manager

HeyPublisher gives you a better way of managing submissions directly within WordPress.

### Description

This plugin allows you to accept submissions from writers without creating accounts for them in WordPress.  You define categories and other filters to ensure you only receive the submissions that meet the needs of your publication.

#### How HeyPublisher Submission Manager Helps

Normally if you wanted to allow readers to submit their writing to your publication you would need to create an account in WordPress for each writer, then educate them on how to write and edit Posts within the publishing tool.

With this plugin that is no longer required.  [HeyPublisher](https://www.heypublisher.com) collects the submissions and this plugin gives you the ability to review the submissions within WordPress.  You no longer have to manage user accounts or worry about unwanted posts filling up your system.

Mark submissions for review, accept submissions for publication, or reject submissions to remove them from your slush pile.  All acceptance/rejection communications with the writer are handled automatically by HeyPublisher using email templates you easily control from within WordPress.

You define the type of writing you want to receive.  You control the flow of content coming into your publication.  You have more time to run your publication.

### Installation

#### Install Plugin

* Go to the [HeyPublisher GitHub repository](https://github.com/HeyPublisher/heypublisher-submission-manager) and download the plugin.

* Click on the Download ZIP link and save it in your `wp-content/plugins` directory.
![Download Zip File](images/install-1.jpg)


* Extract the files to a sub-directory called `heypublisher-submission-manager`.

  *Using cURL, you can do all of the above with one command:*

<!--  `curl -Lo heypub.zip https://github.com/HeyPublisher/heypublisher-submission-manager/archive/master.zip && unzip heypub.zip && mv heypublisher-submission-manager-master heypublisher-submission-manager` -->

* Active the plugin via the WordPress plugin menu.

* You can keep the plugin updated within WordPress by clicking on the link.
![Update Plugin](images/install-4.jpg)


#### Validate Plugin

Once activated you will need to validate the plugin by clicking on the HeyPublisher link in the newly created menu and follow the on-screen instructions.

![Validate Plugin](images/install-2.jpg)

Validation collects some basic information about your publication and yourself as the administrator of this publication:

![Validate Plugin](images/install-3.jpg)

If your publication already exists in the [HeyPublisher.com](https://heypublisher.com/publishers/search) database, enter the name _exactly as it appears_ in our database.  If it doesn't already exist, simply enter the name as you want it to appear within our database.

If you already have an account with HeyPublisher, enter the email address and password you use to login.  If you do not already have an account, enter the email address and password you would like to use.  This username and password will be used if you ever need to reinstall or upgrade the plugin.  This information is also used to ensure that **_ONLY YOU_** can modify your publication's listing in the HeyPublisher database.

#### Configure Plugin

After the plugin has been validated configure it to meet your submission requirements.  All information entered on this screen is used by HeyPublisher to help filter the submissions you receive.  The configuration sections are:

* **Publication Information:** tell writers about your publication.
![Publication Information screen](images/screenshot-2.jpg)

* **Contact Information:** tell writers how to contact you.
![Publication Information screen](images/screenshot-3.jpg)

* **Submission Guidelines:** select the page that contains your submission guidelines (if applicable).

* **Submission Form:** select or create the page that will contain your submission form on your website (if applicable).
![Submission Form Configuration](images/screenshot-4.jpg)

* **Submission Criteria:** tell writers which genres of work you accept - and how these should map to your internal categories when you "accept" a work for publication.  Additionally, you can select whether or not to accept simultaneous submissions and multiple submissions.
![Submission Categories](images/screenshot-5.jpg)

* **Notification Options :** tell HeyPublisher when you want to send notifications to the writer.  You can customize the emails sent to writers in the Response Templates section.
![Writer Notifications](images/screenshot-6.jpg)


<!--
* **Payment Options :** indicate whether or not your publication pays writers for their work.
-->

* **Miscellaneous :** turn this on if your publication uses multibyte characters.
![Miscellaneous](images/screenshot-7.jpg)

Once you have made the appropriate configurations, click the "Update" button.

### Frequently Asked Questions

If you have any questions not answered here, [please email us](mailto:support@heypublisher.com?subject=Question+about+plugin).

* **What happens when I save a submission for "later review"?**
If you do not allow simultaneous submissions, this puts a lock on the work preventing the writer from submitting it to another press while you are considering whether or not to publish it.  The writer, however, may choose to withdraw their submission if it stays in this state for too long.

* **What happens when I accept a submission?**
When you accept a submission a copy of the submission is inserted as a Post into your WordPress system.  The post is marked as 'pending' so you can easily find the accepted submission and make any necessary edits to it prior to publication.

* **What happens when I reject a submission?**
When you reject a submission it is immediately removed from your slush pile and the writer is notified of your decision automatically.  If the work had previously been 'accepted' by you, then rejecting it would also remove it from your pending posts.

* **Can I reject a published submission?**
No - once you publish an accepted submission, the author is automatically notified that their work has been published by you and the work is removed from your "Submissions" administration  screen.

* **We don't have submission guidelines.  Can we still use HeyPublisher?**
Yes - absolutely.  The plugin allows you to either select a Page in your WordPress publication that contains your submission guidelines, or input custom submission guidelines you want displayed to writers.  These guidelines are published on the HeyPublisher website where writers who are interested in submitting to your publication can see them.  

* **Can we change the content of the emails sent to writers?**
Yes - you can define custom response templates that contain whatever message you want to send to your writers.  These emails are sent automatically whenever you reject, accept, or save a submission for later review.  An email is also sent the first time an editor reads a new submission.  Click on **Response Templates** in the side-bar and follow the on-screen instructions.

### Managing the Slushpile
#### Open Submissions

From this screen you can view and manage the submissions received by your publication.  Clicking on the plus icon will display the author's bio, if they have provided one.  Clicking on the title of the submission will allow you to 'preview' the submission.
![Open Submissions](screenshot-8.jpg)

#### Submission Status

When 'previewing' a submission, the submission status will show up on the right side of the screen, giving you a quick overview of the state of the submission.  This includes information about when the submission was submitted, when it was last viewed, and allows you to accept or reject the submission.  Any note you add to the submission here will be sent to the author if you have **Writer Notifications** turned on.

![Submission Status](screenshot-9.jpg)

#### Reimport

If for some reason you need a writer to submit an update to their submission after you have already accepted it, simply go to the submission in the Open Submissions screen and select **Reimport Into WordPress**.  Once you click the "Update Submission" button the new version will be imported, overwriting the previous version of the Post.

![Reimport](screenshot-11.jpg)

### Statistics

The homepage of the plugin will give you a quick snapshot of your slushpile.
![Statistics screen](screenshot-10.jpg)


CHANGELOG.md
## Changelog
+ Minimum plugin version supported by HeyPublisher service: 2.8.0
+ Always use the latest version of the plugin.


### 3.3.0
+ Released: 2021-09-10
+ [[#35](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/35)] : Normalize all calls to make a URL into one function, including:
  + Removed `form_action` function and replaced all calls to it to use the the newer `nonced_url` function as it's more flexible.
+ [[#63](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/63)] : Update Submission State diagram.
+ [[#64](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/64)] : Remove unnecessary code.
+ [[#97](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/97)] : Include category on submission detail screen.
+ [[#121](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/121)] : Error page when trying to access submission with unusual file extension.
+ [[#122](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/122)] : Add option to suppress Author Info when reviewing submissions.
+ [[#124](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/124)] : Validate against [WordPress Version 5.8](https://wordpress.org/support/wordpress-version/version-5-8/).
+ [[#125](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/125)] : Submission List and Details should now use JSON API endpoints, including:
  + Removed all calls to old XML endpoint `api/v1/publisher/fetch_categories` as it's been replaced by JSON endpoint `api/publishers/:id/genres`
  + Removed all calls to old XML endpoint `api/v1/submissions/fetch_pending_submissions` as it's been replaced by JSON endpoint `api/publishers/:id/submissions`
+ Removed old `Loudlever` class and all references to it.  Relevant code was replaced by `Base` class a while ago.
+ Plugin build date display on statistics page was using the date of upgrade, not date of plugin build, so fixed this.
+ **NOTE:** HeyPublisher support for plugin versions older than 2.8.0 has been removed.  Users of versions older than 2.8.0 will receive an upgrade notice and functionality will be limited.

### 3.2.1
+ Released: 2021-06-26
+ [[#118](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/118)] : jQuery version getting loaded with WordPress no longer returns value of current state of checkbox, instead returning initial state of checkbox.  Fixed to use `$(src).prop('checked')` instead.

### 3.2.0
+ Released: 2021-05-31
+ [[#113](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/113)] : Plugin version is sent to the server with all requests, so appropriate data for that version can be returned.
+ [[#114](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/114)] : Fixing issue where file upload errors so writer knows how to fix the problem.
+ [[#116](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/116)] : Validated functionality through WordPress Version 5.7.2.

### 3.1.2
+ Released: 2021-02-06
+ [[#111](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/111) : Fixed warning that was showing on new install first time options page is displayed.
+ Updated the options page to make use of the new 'accepting_subs' attribute in the JSON endpoint.  This is now being used to toggle the "Accepting Submissions" dropdown.

### 3.1.1
+ Released: 2020-09-25
+ [[#110](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/110) : Fix bug that prevented turning off the submissions form

### 3.1.0
+ Released: 2020-08-30
+ Fixed having the wrong year in the 2020 release notes.
+ Plugin is now based on the HeyPublisher/Base class that other plugins are based upon.
  + Updated all code to ensure there are no conflicts with older versions
+ Validated functionality through [WordPress Version 5.5](https://wordpress.org/support/wordpress-version/version-5-5/).

### 3.0.1
+ Released: 2020-06-27
+ Update to updater to fix issue related to updating from GitHub when plugin was manually updated to 3.0.0.

### 3.0.0
+ Released: 2020-06-25
+ Reworked plugin to leverage new HeyPublisher JSON API.
+ [[#102](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/102) : Publishers can now set custom submission guidelines to be displayed on HeyPublisher.com.  Before, HeyPublisher would pull these guidelines directly from the publisher website, but that sometimes resulted in marketing text getting included when it should not have been.
+ [[#69](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/69) : Introduced a new Config class that handles all configuration parameter settings throughout the code base.
+ [[#61](https://github.com/HeyPublisher/heypublisher-submission-manager/issues/61)] : Better error messaging when HeyPublisher.com is undergoing maintenance.
+ Implemented caching of publication types and genres to speed up loading time.
+ Fixed references to non-secure URLs that were throwing errors in the console log, but did not affect performance.
+ Plugin is now hosted on [GitHub](https://github.com/HeyPublisher/heypublisher-submission-manager).


### 2.8.3
+ Released: 2019-12-06
+ [[#98](https://github.com/aguywithanidea/heypublisher-submission-manager/issues/98)] : Added link to the Statuscake status page, in cases where errors are thrown.
+ Additionally ....
  + Validated functionality through [WordPress Version 5.3](https://wordpress.org/support/wordpress-version/version-5-3/).

### 2.8.2
+ Released: 2018-05-09

heypublisher-sub-mgr.php
<?php
/*
Plugin Name: HeyPublisher Submission Manager
Plugin URI: https://github.com/HeyPublisher/heypublisher-submission-manager
Description: HeyPublisher is a better way of managing submissions for your WordPress-powered publication.
Author: HeyPublisher
Author URI: https://www.heypublisher.com
Version: 3.3.0
Requires at least: 5.0


  Copyright 2010-2014 Loudlever, Inc. (wordpress@loudlever.com)
  Copyright 2014-2018 Richard Luck (https://github.com/aguywithanidea/)
  Copyright 2019-2021 HeyPublisher, LLC (https://www.heypublisher.com/)

  Permission is hereby granted, free of charge, to any person
  obtaining a copy of this software and associated documentation
  files (the "Software"), to deal in the Software without
  restriction, including without limitation the rights to use,
  copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the
  Software is furnished to do so, subject to the following
  conditions:

  The above copyright notice and this permission notice shall be
  included in all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
  EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
  OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
  NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
  HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
  WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
  FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
  OTHER DEALINGS IN THE SOFTWARE.

*/

/*
 Older versions of PHP may not define DIRECTORY_SEPARATOR so define it here,
 just in case.
*/
if(!defined('DIRECTORY_SEPARATOR')) {
  define('DIRECTORY_SEPARATOR','/');
}
/**
*  DEFINITIONS
*/
define('HEY_DIR', dirname(plugin_basename(__FILE__)));


/*
---------------------------------------------------------------------------------
  OPTION SETTINGS

  1.1.0 => 29
  1.2.0 => 35
  1.2.4 => 38
  1.3.0 => 40
  1.4.1 => 45
  1.4.2 => 48
  1.4.3 => 49
  1.4.4 => 50
  1.4.5 => 51
  1.5.0 => 52
  1.5.1 => 53
  2.0.0 => 60
  2.0.1 => 61
  2.1.0 => 62
  2.2.0 => 63
  2.3.0 => 64
  2.4.0 => 65
  2.5.0 => 66
  2.6.0 => 67
  2.6.1 => 68
  2.6.2 => 69
  2.6.3 => 70
  2.7.0 => 71
  2.8.0 => 72
  2.8.1 => 73
  2.8.2 => 74
  2.8.3 => 75
  2.9.0 => 76
  3.0.0 => 80
  3.0.1 => 81
  3.1.0 => 82
  3.1.1 => 83
  3.1.2 => 84
  3.2.0 => 88
  3.2.1 => 89
  3.3.0 => 90 => min server req: 4.0.3
---------------------------------------------------------------------------------
*/

// Configs specific to the plugin
// Build Number (must be a integer)
define('HEY_BASE_URL', get_option('siteurl').'/wp-content/plugins/'.HEY_DIR.'/');
define("HEYPUB_PLUGIN_BUILD_DATE", "2021-09-01");
// Version Number (can be text)
define("HEYPUB_PLUGIN_BUILD_NUMBER", "90");  // This controls whether or not we get upgrade prompt
define("HEYPUB_PLUGIN_VERSION", "3.3.0");
define("HEYPUB_PLUGIN_TESTED", "5.8.0");

# Base domain
$domain = 'https://www.heypublisher.com';
$debug = (getenv('HEYPUB_DEBUG') === 'true');
if ($debug) {
  $domain = 'http://127.0.0.1:3000';
}
define('HEYPUB_PLUGIN_ERROR_CONTACT','support@heypublisher.com');

define('HEYPUB_PLUGIN_FULLPATH', dirname(__FILE__));
// How to connect to the service
define('HEYPUB_FEEDBACK_EMAIL_VALUE','support@heypublisher.com?subject=HeyPublisher%20Wordpress%20Plugin');
define('HEYPUB_FEEDBACK_GETSATISFACTION','mailto:support@heypublisher.com?Subject=HeyPublisher%20Submission%20Manager');
// define('HEYPUB_SVC_URL_STYLE_GUIDE','https://blog.heypublisher.com/docs/plugins/wordpress/style_guide/');     # designates the URL of the style guide
define('HEYPUB_DOMAIN', $domain);
# Stylesheet for plugin resides on HP server now
define('HEYPUB_SVC_STYLESHEET_URL',HEYPUB_DOMAIN . '/stylesheets/wordpress/plugin.css?' . HEYPUB_PLUGIN_VERSION);

// TODO: remove these defines and reference directly from XML code that makes query
define('HEYPUB_SVC_URL_AUTHENTICATE','publishers/fetch_or_create');           # initial plugin authentication

# if this changes, plugin will not work.  You have been warned
// TODO: tie token value to version / host
define('HEYPUB_SVC_TOKEN_VALUE','534ba1c699ca9310d7acf4832e12bed87c4d5917c5063c58382e9766bca11800');

// Locally stored in database
// install key contains info about the install iteself
// config key contains cached responses from server for faster page loads
define('HEYPUB_PLUGIN_OPT_INSTALL', '_heypub_plugin_opt_install');
define('HEYPUB_PLUGIN_OPT_CONFIG', '_heypub_plugin_options');

define('HEYPUB_SVC_URL','_heypub_service_url');

// messages for sending to the User
// define('HEYPUB_OPT_MSG_REJECT','_heypub_opt_msg_reject');     # Text of rejection notice

// Info about the Page that will be created to house submissions, if needed.
define('HEYPUB_SUBMISSION_PAGE_TITLE','Submission Form');
define('HEYPUB_SUBMISSION_PAGE_REPLACER','[__HEYPUBLISHER_SUBMISSION_FORM_GOES_HERE__]');

// These keys are stored in the usermeta table and are NOT 'deleted' when we uninstall the plugin
// This way we can always track which users/posts have been affected by HP plugin
define('HEYPUB_USER_META_KEY_AUTHOR_ID','_heypub_user_meta_key_author_id');
define('HEYPUB_USER_META_KEY_AUTHOR_OID','_heypub_user_meta_key_author_oid');
define('HEYPUB_POST_META_KEY_SUB_ID','_heypub_post_meta_key_sub_id');

/**
* Load all of the plugin files
*/
global $hp_xml, $hp_base, $hp_config;
global $hp_updater;

// Updater needed to check for plugin updates now that we're hosted on GitHub
if (!class_exists("\HeyPublisher\Base\Updater")) {
  require_once(HEYPUB_PLUGIN_FULLPATH . '/include/classes/HeyPublisher/Base/Updater.class.php');
}
// initialize the updater and test for update
$hp_updater = new \HeyPublisher\Base\Updater( __FILE__ );
$hp_updater->set_repository( 'heypublisher-submission-manager' ); // set repo
$hp_updater->initialize(HEYPUB_PLUGIN_TESTED); // initialize the updater

// This class fetches data from and stores data in WP db
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/classes/HeyPublisher/Config.class.php');
// XML api v1
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/HeyPublisherXML/HeyPublisherXML.class.php');
// dunno ?!
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/HeyPublisher/HeyPublisher.class.php');
$hp_config = new \HeyPublisher\Config;
$hp_xml = new HeyPublisherXML;
$hp_base = new HeyPublisher;
$hp_xml->debug = $debug;
$hp_xml->log("Loading plugin::");
$hp_xml->log(sprintf("\n HEYPUB_PLUGIN_FULLPATH = %s",HEYPUB_PLUGIN_FULLPATH));
// These files are required for basic functions
require_once(HEYPUB_PLUGIN_FULLPATH.'/include/heypub-template-functions.php');

// Load the classes
// Main page
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/classes/HeyPublisher/Page/Overview.class.php');
$hp_main = new \HeyPublisher\Page\Overview;

// Plugin configuration options
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/classes/HeyPublisher/Page/Options.class.php');
$hp_opt = new \HeyPublisher\Page\Options;

// Submissions
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/classes/HeyPublisher/Page/Submissions.class.php');
$hp_subs = new \HeyPublisher\Page\Submissions;

// Email Templates
load_template(HEYPUB_PLUGIN_FULLPATH . '/include/classes/HeyPublisher/Page/Email.class.php');
$hp_email = new \HeyPublisher\Page\Email;

// Initiate the callbacks
register_activation_hook (__FILE__, 'heypub_init');
register_deactivation_hook( __FILE__, 'heypub_uninit');
// Register the adminstration menu
add_action('admin_init', 'RegisterHeyPublisherAdminStyle');
add_action('admin_menu', 'RegisterHeyPublisherAdminMenu');

// IMPORTANT: If you have custom posts and want to have HeyPublisher send a notice
// to the writer when the submission is either published, or rejected after acceptance,
// you will need to modify the 3 add_action() statements below.
// Simply change '_post' to '_your_custom_post type'
// For example, if your custom post type is called 'story',
// Change
//          add_action('publish_post',array($hp_subs,'publish_post'));
// to read:
//          add_action('publish_story',array($hp_subs,'publish_post'));

// Marks the submission as 'published' in HeyPublisher and removes it from your Submission Summary screen:
add_action('publish_post',array($hp_subs,'publish_post'));
// Marks a previously accepted submission as 'rejected' in HeyPublisher and removes it from your Submission Summary screen:
// this one executes when the submission is moved to 'trash'
add_action('trash_post',array($hp_subs,'delete_post_cleanup'));
// this one executes when you skip the trash an simply 'delete' the submission
add_action('delete_post',array($hp_subs,'delete_post_cleanup'));

// Ensure the post edit screen shows pre-acceptance actions in the 'revisions' metabox
add_action( 'do_meta_boxes', array($hp_subs,'revisions_meta_box'));

// Ensure we go through the upgrade path even if the user simply installs
// a new version of the plugin over top of the old plugin.
if ($hp_config->install['version_current'] != HEYPUB_PLUGIN_BUILD_NUMBER) {
  heypub_init();
}

// register the admin styles
function RegisterHeyPublisherAdminStyle() {
  wp_register_style( 'heypublisher', plugins_url( 'include/css/heypublisher.css', __FILE__ ), array(), HEYPUB_PLUGIN_VERSION );
}

/**
*  Configure and Register the Admin Menu
*  Invoke the hook, sending function name
*/
function RegisterHeyPublisherAdminMenu(){
  global $hp_xml, $hp_opt, $hp_subs, $hp_main, $hp_email, $hp_config;
  // Initilise the plugin for the first time here. This gets called when you click the HeyPublisher link.
  $admin_menu = add_menu_page('HeyPublisher Stats','HeyPublisher', 8, HEY_DIR, array($hp_main,'page_prep'), 'dashicons-book-alt');
  add_action("admin_print_styles-$admin_menu", 'HeyPublisherAdminHeader' );

  // Configure Options
  $admin_opts = add_submenu_page( HEY_DIR , 'Configure HeyPublisher', 'Plugin Options', 'manage_options', $hp_opt->slug, array($hp_opt,'options'));
  // add_action("load-$admin_opts", array($hp_opt,'help_menu') );
  add_action("admin_print_styles-$admin_opts", 'HeyPublisherAdminHeader' );
  add_action("admin_print_scripts-$admin_opts", 'HeyPublisherAdminInit');

  if ($hp_config->is_validated) {
    // Submission Queue
    $admin_sub = add_submenu_page(HEY_DIR , 'HeyPublisher Submissions', 'Submissions', 'edit_others_posts', $hp_subs->slug, array($hp_subs,'action_handler'));
    add_action("admin_print_styles-$admin_sub", 'HeyPublisherAdminHeader' );
    add_action("admin_print_scripts-$admin_sub", 'HeyPublisherAdminInit');

    // Response Templates
    $admin_temps = add_submenu_page( HEY_DIR , 'HeyPublisher Email Templates', 'Email Templates', 'manage_options', $hp_email->slug, array($hp_email,'action_handler'));
    add_action("load-$admin_temps", array($hp_email,'help_menu') );
    add_action("admin_print_styles-$admin_temps", 'HeyPublisherAdminHeader' );
    add_action("admin_print_scripts-$admin_temps", 'HeyPublisherAdminInit');
  }
  // Uninstall Plugin - now moved to Overview page :)
}

// Load the custom .css that's already been registered
function HeyPublisherAdminHeader() {
  wp_enqueue_style( 'heypublisher' );
}
// Load the custom .js
function HeyPublisherAdminInit() {
	$parts = array(WP_PLUGIN_URL,HEY_DIR,'include','js','heypublisher.js');
	$url = implode('/',$parts);
  wp_enqueue_script('heypublisher', $url, array('jquery'), HEYPUB_PLUGIN_VERSION );
  // wp_enqueue_style('heypub_font_css', '//maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css', array(), HEYPUB_PLUGIN_VERSION);
}

// /*
// Handler for display the Response Templates admin page
// */
// function heypub_response_templates() {
// require_once(HEYPUB_PLUGIN_FULLPATH.'/include/HeyPublisher/HeyPublisherResponse.class.php');
//   $hp_res = new HeyPublisherResponse;
//   $hp_res->handler();
// }

// -----------------------------------------------------------------------------
// Initialize / Upgrade
// -----------------------------------------------------------------------------
function heypub_init(){
  global $hp_xml, $hp_config;
  $hp_xml->log("heypub_init()");

  // NEW Install Path
  if (get_option(HEYPUB_PLUGIN_OPT_INSTALL) == false) {
    $hp_config->initialize();
    if (function_exists('get_bloginfo')) {
      $hp_config->set_config_option('name',get_bloginfo('name'));
      // TODO: these need to change
      $hp_config->set_config_option('url',get_bloginfo('url'));
      $hp_config->set_config_option('editor_email',get_bloginfo('admin_email'));
      if (function_exists('get_feed_permastruct')) {
        $hp_config->set_config_option('rss',get_bloginfo('rss2_url'));
      }
    }
    $hp_config->set_config_option('editor_name','Editor');
  }

  // now check for a normal upgrade path
  $opts = $hp_config->install;
  if ($opts['version_current'] != HEYPUB_PLUGIN_BUILD_NUMBER) {
    $hp_config->logger->debug("\tupgrading plugin");
    // this is the 'normal' upgrade path.
    if ($opts['version_current'] < 40) {  // upgrade to 1.3.0 options
      $hp_config->logger->debug("\tupgrading to 40");
      $hp_config->set_config_option('notify_submitted','1');
      $hp_config->set_config_option('notify_read','1');
      $hp_config->set_config_option('notify_rejected','1');
      $hp_config->set_config_option('notify_published','1');
      $hp_config->set_config_option('notify_accepted','1');
      $hp_config->set_config_option('notify_under_consideration','1');
      $hp_config->set_config_option('reprint_subs','0');
      if (function_exists('get_bloginfo')) {
        // for feed info, also need to test for 'get_feed_permastruct()')
        if (function_exists('get_feed_permastruct')) {
          $hp_config->set_config_option('rss',get_bloginfo('rss2_url'));
        }
      }
    }
    // not all of our 1.3.1 version users have the bloginfo for rss set, upgrade them
    if ($opts['version_current'] < 42) {  // upgrade to 1.3.2 options
      $hp_config->logger->debug("\tupgrading to 42");
      if (function_exists('get_bloginfo') && function_exists('get_feed_permastruct')) {
          $hp_config->set_config_option('rss',get_bloginfo('rss2_url'));
      }
    }
    // ensure 2.3.0 versons have base config for mailchimp, even if not used
    if ($opts['version_current'] < 64) {  // upgrade to 2.3.0 options
      $hp_config->logger->debug("\tupgrading to 64");
      $hp_config->set_config_option('mailchimp_active', false);
      $hp_config->set_config_option('mailchimp_api_key', null);
      $hp_config->set_config_option('mailchimp_list_id', null);
    }
    if ($opts['version_current'] < 65) {  // upgrade to 2.4.0 options
      $hp_config->logger->debug("\tupgrading to 65");
      $hp_config->set_config_option('notify_withdrawn','1');
    }
    // Upgraded to the 3.0.0 vesion of options
    if ($opts['version_current'] < 80) {
      $hp_config->logger->debug("\tupgrading to 80");
      // need to migrate keys
      $address = array(
        'street'  => $hp_config->get_config_option('address'),
        'city'    => $hp_config->get_config_option('city'),
        'state'   => $hp_config->get_config_option('state'),
        'country'   => $hp_config->get_config_option('country'),
        'zipcode'   => $hp_config->get_config_option('zipcode')
      );
      $hp_config->set_config_option('address',$address);
      // we no longer need these keys set in config
      $hp_config->kill_config_option('city');
      $hp_config->kill_config_option('state');
      $hp_config->kill_config_option('zipcode');
      $hp_config->kill_config_option('country');
      $notifications = array(
        'read'        => $hp_config->get_config_option('notify_read'),
        'considered'  => $hp_config->get_config_option('notify_under_consideration'),
        'accepted'    => $hp_config->get_config_option('notify_accepted'),
        'rejected'    => $hp_config->get_config_option('notify_rejected'),
        'published'   => $hp_config->get_config_option('notify_published'),
        'withdrawn'   => $hp_config->get_config_option('notify_withdrawn'),
      );
      $hp_config->set_config_option('notifications',$notifications);
      // we no longer need these keys set in config
      $hp_config->kill_config_option('notify_submitted');
      $hp_config->kill_config_option('notify_read');
      $hp_config->kill_config_option('notify_rejected');
      $hp_config->kill_config_option('notify_accepted');
      $hp_config->kill_config_option('notify_published');
      $hp_config->kill_config_option('notify_under_consideration');
      $hp_config->kill_config_option('notify_withdrawn');

      // TODO: Migrate the 'accepts'
      $accepts = array(
        'reprints'      => $hp_config->get_config_option('reprint_subs'),
        'simultaneous'  => $hp_config->get_config_option('simu_subs'),
        'email'         => false,  # we didn't have a default for this
        'multiple'      => $hp_config->get_config_option('multi_subs'),
        'multibyte'     => $hp_config->get_config_option('turn_off_tidy')
      );
      $hp_config->set_config_option('accepts',$accepts);
      $hp_config->kill_config_option('reprint_subs');
      $hp_config->kill_config_option('simu_subs');
      $hp_config->kill_config_option('multi_subs');
      $hp_config->kill_config_option('turn_off_tidy');

      //  Need to migrate from $saved_genres = $this->xml->get_category_mapping(); to remote mapping
      // $hp_config->set_config_option('category_map',$hp_config->get_config_option('categories'));
      $map = $hp_xml->get_category_mapping();
      $hp_xml->log(sprintf("heypub_init() \$map = \n\t%s",print_r($map,1)));
    }
    // Upgrade to 3.3.0 options
    if ($opts['version_current'] < 90) {
      $hp_config->logger->debug("\tupgrading options to build 90");
      $author = array(
        'blind_review'  => false,
        'sync_bio'      => false
      );
      $hp_config->set_config_option('author',$author);
    }

    // finally - ensure that the last version and current version are set
    $data = array(
      'version_last'      => $opts['version_current'],
      'version_last_date' => $opts['version_current_date'],
      'version_current'   => HEYPUB_PLUGIN_BUILD_NUMBER,
      'version_current_date' => date('Y-m-d')
    );
    $hp_config->set_install_options($data);
  }
  // otherwise we keep chugging along
}
// TODO: This needs to be fixed for new API support
function heypub_uninit() {
  global $hp_xml, $hp_config, $hp_opt;
  $opts = $hp_config->config;
  $install = $hp_config->install;
  if ($install != FALSE && isset($install['user_oid']) ) {
    // disable the publisher in the db
    $opts['accepting_subs'] = false;
    $opts['genres_list'] = false;
    $opts['guide'] = get_permalink($opts['sub_guide_id']);
    // TODO: This should change to api->update_publisher();
    $hp_opt->pubapi->deactivate();
  }
  $hp_config->install = false;
  $hp_config->config = false;
  delete_option(HEYPUB_PLUGIN_OPT_INSTALL);
  delete_option(HEYPUB_PLUGIN_OPT_CONFIG);
  return;
}



?>
